File: src/07-fibonacci/rust/src/main.rs
/*!
 * Polyglot Programming
 *
 * Copyright (c) 2024 Adam Burucs
 *
 * MIT Licensed
 */

fn fibonacci(n: u32) -> Result<Vec<u32>, String> {
    let mut fib_numbers = vec![1, 1];
    let _prev_prev_index = 0;
    match n {
        0 => {
            return Err(String::from(
                "Error: Wrong input number, it should be greater or equal than 1.",
            ))
        }
        1 | 2 => return Ok(fib_numbers),
        3_u32..=u32::MAX => {
            for (prev_prev_index, _i) in (3..n).enumerate() {
                fib_numbers.push(fib_numbers[prev_prev_index] + fib_numbers[prev_prev_index + 1]);
            }
        }
    }
    Ok(fib_numbers)
}

fn main() {
    println!("Fibonacci sequence in Rust");
    let fib_numbers = fibonacci(10);
    match fib_numbers {
        Ok(val) => {
            let s = format!("{val:?}");
            println!("{s}");
        }
        Err(e) => println!("{e}"),
    }
}

File: src/07-fibonacci/typescript/src/main.ts
/*!
 * Polyglot Programming
 *
 * Copyright (c) 2024- Adam Burucs
 *
 * MIT Licensed
 */

function fibonacci(n: number): Result<number[], string> {
  let fibNumbers = [1, 1];
  let prevPrevIndex = 0;
  if (n < 1) {
    return {
      ok: false,
      err: "Error: Wrong input number, it should be greater or equal than 1.",
    };
  }
  if (n === 1 || n === 2) {
    return { ok: true, val: fibNumbers };
  }
  for (let i = 3; i <= n; i++) {
    fibNumbers.push(fibNumbers[prevPrevIndex] + fibNumbers[prevPrevIndex + 1]);
    prevPrevIndex++;
  }
  return { ok: true, val: fibNumbers };
}

interface Result<T, E> {
  ok: boolean;
  val?: T;
  err?: E;
}

function main() {
  console.log("Fibonacci sequence in TypeScript");
  const fibNumbers = fibonacci(10);
  if (fibNumbers.ok) {
    console.log(fibNumbers.val);
  } else {
    console.log(fibNumbers.err);
  }
}

main();
